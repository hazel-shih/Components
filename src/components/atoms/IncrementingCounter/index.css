.incrementing-counter {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;

	& > img {
		width: 50px;
		height: 50px;
		margin-bottom: 15px;
	}

	& > .counter {
		color: var(--neutral-n1-primary-text);
		font-weight: 500;
		margin-bottom: 5px;
		font-size: 48px;
		letter-spacing: 3px;
	}

	& > p {
		color: var(--neutral-n1-primary-text);
		font-size: 18px;
	}
}
