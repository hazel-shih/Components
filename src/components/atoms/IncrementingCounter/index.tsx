import React, { useState, useEffect } from 'react';
import classnames from 'classnames';

import styles from './index.css';

interface IncrementingCounterProperty {
	imgSrc: string;
	targetNumber: number;
	label: string;
	className?: string;
}

const numberWithCommas = (number: number) =>
	// 待弄懂
	number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

const IncrementingCounter: React.FC<IncrementingCounterProperty> = ({
	imgSrc,
	targetNumber,
	label,
	className,
}) => {
	// 300 is custom speed, the bigger the slower
	const increment = targetNumber / 300;
	const [currentNum, setCurrentNum] = useState<number>(0);

	useEffect(() => {
		if (currentNum < targetNumber) {
			setTimeout(() => {
				setCurrentNum(Math.ceil(currentNum + increment));
			}, 1);
		}
		if (currentNum >= targetNumber) {
			setCurrentNum(targetNumber);
		}
	}, [currentNum]);

	return (
		<div className={classnames(styles.incrementingCounter, className)}>
			<img src={imgSrc} alt="brand-logo" />
			<h1 className={styles.counter}>{numberWithCommas(currentNum)}</h1>
			<p>{label}</p>
		</div>
	);
};

export default IncrementingCounter;
