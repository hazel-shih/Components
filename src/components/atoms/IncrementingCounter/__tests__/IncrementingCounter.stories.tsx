import React from 'react';
import { storiesOf } from '@storybook/react';

import IncrementingCounter from 'components/atoms/IncrementingCounter';
import Instagram from 'images/icon/instagram.png';

const stories = storiesOf('atoms/IncrementingCounter', module);

stories.add('__interactive', () => (
	<IncrementingCounter imgSrc={Instagram} targetNumber={1234567} label="Instagtram Followers" />
));
