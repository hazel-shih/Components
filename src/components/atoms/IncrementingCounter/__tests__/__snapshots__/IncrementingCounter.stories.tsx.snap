// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="incrementingCounter"
  >
    <img
      alt="brand-logo"
      src="test-file-stub"
    />
    <h1
      class="counter"
    >
      0
    </h1>
    <p>
      Instagtram Followers
    </p>
  </div>
</DocumentFragment>
`;
